Add support for cloud sync, specifically GCP (#3223)
GothenburgBitFactory/taskwarrior: taskchampion/docs/src/sync-protocol.md

# Server-Replica Protocol

The server-replica protocol is defined abstractly in terms of request/response transactions.

The protocol builds on the model presented in the previous chapters, and in particular on the synchronization process.

## Clients

From the protocol's perspective, replicas accessing the same task history are indistinguishable, so this protocol uses the term "client" to refer generically to all replicas replicating a single task history.

## Server

A server implements the requests and responses described below.
Where the logic is implemented depends on the specific implementation of the protocol.

For each client, the server is responsible for storing the task history, in the form of a branch-free sequence of versions.
It also stores the latest snapshot, if any exists.
From the server's perspective, snapshots and versions are opaque byte sequences.

## Version Invariant

The following invariant must always hold:

> > All versions accessible from the server, when linked by parent-child relationships, must form a single chain.

Reviewer: Maybe this is more clear? I find "accessible" here ambiguous and confusing.

```suggestion
> All versions stored on the server, when linked by parent-child relationships, must form a single chain.
```

Author: I was leaving room open for extra versions that are created on the server but then fail the "latest" update. Those are stored on the server, but not accessible via the API. Maybe there's another, longer but still succinct way to phrase that?

Reviewer: Ok, with that context in mind it makes sense to me now. Maybe

> With respect to versions replicated by the server, the following invariant holds:
> All versions are linked by parent-child relationships to form a single chain.

If you don't think that's an improvement then forget it. Either I'm the only one who struggles with it or somebody else will come along with a better idea.

Author: That's great -- thanks!

> That is, each version must have no more than one parent and one child, and no more than one version may have zero parents or zero children.

## Data Formats

> All data sent to the server is encrypted by the client, using the scheme described in the "Encryption" chapter.

Reviewer: "All data" is over-promising because some metadata is not encrypted.

```suggestion
Data sent to the server is encrypted by the client, using the scheme described in the "Encryption" chapter.
```

Author: I went with "Task data  .."
